source/lib/cache.js

```diff
@@ -1,97 +1,147 @@
-	if (!stored.timeToCache) {
-		return {isCached: true, isExpired: false, value: stored.value}
-	}
+/// MARK: Utilities
 
-	const date = new Date(stored.dateCached)
-	const isExpired = needsUpdate(date, stored.timeToCache)
-	return {isCached: true, isExpired, value: stored.value}
+async function setItem(
+	key: string,
+	value: any,
+	{ttl, source}: {ttl?: [number, string], source?: string},
+) {
+	console.log('setItem', {key, value, ttl, source})
+	await AsyncStorage.multiSet([
+		[`fp:${key}:data`, JSON.stringify(value)],
+		[`fp:${key}:ts`, new Date().toUTCString()],
+		[`fp:${key}:ttl`, JSON.stringify(ttl)],
+		[`fp:${key}:source`, source],
+	])
 }
 
-/// MARK: Utilities
+async function getItem(key: string): CacheResult<any> {
+	console.log('getItem', key)
+
+	let [[, value], [, ts], [, ttl], [, source]] = await AsyncStorage.multiGet([
+		`fp:${key}:data`,
+		`fp:${key}:ts`,
+		`fp:${key}:ttl`,
+		`fp:${key}:source`,
+	])
 
-function setItem(key: string, value: any, cacheTime?: [number, string]) {
-	const dataToStore = {
-		dateCached: new Date().toUTCString(),
-		timeToCache: cacheTime,
-		value: value,
+	return {
+		value: JSON.parse(value),
+		source,
+		...annotateCacheEntry({ts, ttl}),
 	}
-	return AsyncStorage.setItem(`aao:${key}`, JSON.stringify(dataToStore))
-}
-function getItem(key: string): CacheResultType<any> {
-	return AsyncStorage.getItem(`aao:${key}`).then(stored =>
-		annotateCacheEntry(JSON.parse(stored)),
-	)
 }
 
-/// MARK: Help tools
+global._setItem = setItem
+global._getItem = getItem
+global._fetchAndCacheItem = fetchAndCacheItem
 
-const helpToolsKey = 'help:tools'
-const helpToolsCacheTime = [1, 'hour']
-import {type ToolOptions} from '../views/help/types'
-const {data: helpData} = require('../../docs/help.json')
-export function setHelpTools(tools: Array<ToolOptions>) {
-	return setItem(helpToolsKey, tools, helpToolsCacheTime)
-}
-export function getHelpTools(): CacheResultType<?Array<ToolOptions>> {
-	return getItem(helpToolsKey)
-}
-function fetchHelpToolsBundled(): Promise<Array<ToolOptions>> {
-	return Promise.resolve(helpData)
-}
-function fetchHelpToolsRemote(): Promise<{data: Array<ToolOptions>}> {
-	return fetchJson(API('/tools/help'))
-}
-export async function fetchHelpTools(): Promise<Array<ToolOptions>> {
-	const cachedValue = await getHelpTools()
+export async function fetchAndCacheItem(args: {
+	key: string,
+	url?: string,
+	cbForBundledData?: () => Promise<mixed>,
+	ttl?: [number, string],
+	afterFetch?: (parsed: any) => any,
+	force?: boolean,
+	delay?: boolean,
+}): CacheResult<any> {
+	let {
+		key,
+		url,
+		cbForBundledData,
+		ttl,
+		afterFetch,
+		force: shouldForce,
+		delay: shouldDelay,
+	} = args
+
+	let start = Date.now()
 
-	if (process.env.NODE_ENV === 'development') {
-		return fetchHelpToolsBundled()
+	console.log('fetchAndCacheItem', {key, url, cbForBundledData, ttl})
+
+	// if (process.env.NODE_ENV === 'development' && cbForBundledData) {
```

Comment:
Dead code now?

Reply:
Only dead until we're ready to ship the feature.

I like to have this in so that we run off of local data when developing… saves several minutes of head-scratching about "why aren't my changes showing up???" whenever you're doing that.

Reply:
Yep, that's all good.